content/blog/autoscaling-custom-k8s-metric/index.md

```diff
@@ -0,0 +1,243 @@
+
+* In `ListAllMetrics`, the custom metric server defines a custom metric, `px-http-requests-per-second`, on the Pod resource type.
+* The custom metric server queries Pixie's [API](https://docs.px.dev/reference/api/overview) every 30 seconds in order to generate the metric values (HTTP requests/second, by pod).
+* These values can be fetched by subsequent calls to `GetMetricByName` and `GetMetricBySelector`.
+* The server caches the results of the query to avoid unnecessary recomputation every time a metric is fetched.
+
+The custom metrics server contains a [hard-coded](https://github.com/pixie-io/pixie-demos/blob/main/custom-k8s-metrics-demo/pixie-http-metric-provider.go#L33-L48) PxL script (Pixie's [query language](https://docs.px.dev/reference/pxl/)) in order to compute HTTP requests/second by pod. PxL is very flexible, so we could easily extend this script to generate other metrics instead (latency by pod, requests/second in a different protocol like SQL, function latency, etc). 
+
+It's important to generate a custom metric for every one of your pods, because the Kubernetes autoscaler will not assume a zero-value for a pod without a metric associated. One early bug our implementation had was omitting the metric for pods that didn't have any HTTP requests.
+
+## Testing and Tuning
+
+We can sanity check our custom metric via the `kubectl` API:
+
+```bash
+kubectl [-n <ns>] get --raw "/apis/custom.metrics.k8s.io/v1beta2/namespaces/default/pods/*/px-http-requests-per-second"
+```
+
+Let's try it on a demo application, a simple echo server. The echo server has a horizontal pod autoscaler that looks like this:
+
+```yaml
+apiVersion: autoscaling/v2beta2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: echo-service-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: echo-service
+  minReplicas: 1
+  maxReplicas: 10
+  metrics:
+    - type: Pods
+      pods:
+        metric:
+          name: px-http-requests-per-second
+        target:
+          type: AverageValue
+          averageValue: 20
+```
+
+**This autoscaler will try to meet the following goal: 20 requests per second (on average), per pod.** If there are more requests per second, it will increase the number of pods, and if there are fewer, it will decrease the number of pods. The `maxReplicas` property prevents the autoscaler from provisioning more than 10 pods.
+
+We can use [hey](https://github.com/rakyll/hey) to generate artificial HTTP load. This generates 10,000 requests to the `/ping` endpoint in our server.
+
+```bash
+hey -n 10000 http://<custom-metric-server-external-ip>/ping
+```
+
+Let's see what happens to the number of pods over time in our service. In the chart below, the 10,000 requests were generated at second 0. The requests were fast -- all 10,000 completed within a few seconds.
+
+::: div image-xl
+<svg title="The number of pods increases in response to load, then decreases." src='autoscaling-chart-light.png' />
```

Comment:
if you drop the `-light` and instead set `src` to `autoscaling-chart.png`, this will follow light mode/dark mode.

(Note that you will have to make sure all 3 versions of the image file exist, this one is fine)